ec2/volumes_test.go

```diff
@@ -0,0 +1,296 @@
+
+	ids := []string{"vol-1a2b3c4d"}
+	resp, err := s.ec2.Volumes(ids, nil)
+	req := testServer.WaitRequest()
+
+	c.Assert(req.Form["Action"], DeepEquals, []string{"DescribeVolumes"})
+	c.Assert(req.Form["VolumeId.1"], DeepEquals, []string{ids[0]})
+
+	c.Assert(err, IsNil)
+	c.Assert(resp.RequestId, Equals, "59dbff89-35bd-4eac-99ed-be587EXAMPLE")
+	c.Check(resp.Volumes, HasLen, 1)
+	volume := resp.Volumes[0]
+	c.Check(volume.Id, Equals, "vol-1a2b3c4d")
+	c.Check(volume.AvailZone, Equals, "us-east-1a")
+	c.Check(volume.Status, Equals, "in-use")
+	c.Check(volume.VolumeType, Equals, "standard")
+	c.Check(volume.Size, Equals, 80)
+	c.Check(volume.IOPS, Equals, int64(3000))
+	c.Check(volume.Encrypted, Equals, true)
+	c.Check(volume.Tags, HasLen, 0)
+	attachments := volume.Attachments
+	c.Check(attachments, HasLen, 1)
+	attachment := attachments[0]
+	c.Check(attachment.VolumeId, Equals, ids[0])
+	c.Check(attachment.Status, Equals, "attached")
+	c.Check(attachment.Device, Equals, "/dev/sdh")
+	c.Check(attachment.DeleteOnTermination, Equals, false)
+}
+
+func (s *S) TestAttachVolumeExample(c *C) {
+	testServer.Response(200, nil, AttachVolumeExample)
+
+	resp, err := s.ec2.AttachVolume("volume-id", "instance-id", "device")
+	req := testServer.WaitRequest()
+
+	c.Assert(req.Form["Action"], DeepEquals, []string{"AttachVolume"})
+	c.Assert(req.Form["VolumeId"], DeepEquals, []string{"volume-id"})
+	c.Assert(req.Form["InstanceId"], DeepEquals, []string{"instance-id"})
+	c.Assert(req.Form["Device"], DeepEquals, []string{"device"})
+
+	c.Assert(err, IsNil)
+	c.Assert(resp.RequestId, Equals, "59dbff89-35bd-4eac-99ed-be587EXAMPLE")
+	c.Check(resp.VolumeId, Equals, "vol-1a2b3c4d")
+	c.Check(resp.Device, Equals, "/dev/sdh")
+	c.Check(resp.InstanceId, Equals, "i-1a2b3c4d")
+	c.Check(resp.Status, Equals, "attaching")
+}
+
+func (s *S) TestDetachVolumeExample(c *C) {
+	testServer.Response(200, nil, DetachVolumeExample)
+
+	resp, err := s.ec2.DetachVolume("volume-id", "instance-id", "device", true)
+	req := testServer.WaitRequest()
+
+	c.Assert(req.Form["Action"], DeepEquals, []string{"DetachVolume"})
+	c.Assert(req.Form["VolumeId"], DeepEquals, []string{"volume-id"})
+	c.Assert(req.Form["InstanceId"], DeepEquals, []string{"instance-id"})
+	c.Assert(req.Form["Device"], DeepEquals, []string{"device"})
+	c.Assert(req.Form["Force"], DeepEquals, []string{"true"})
+
+	c.Assert(err, IsNil)
+	c.Assert(resp.RequestId, Equals, "59dbff89-35bd-4eac-99ed-be587EXAMPLE")
+	c.Check(resp.VolumeId, Equals, "vol-1a2b3c4d")
+	c.Check(resp.Device, Equals, "/dev/sdh")
+	c.Check(resp.InstanceId, Equals, "i-1a2b3c4d")
+	c.Check(resp.Status, Equals, "detaching")
+}
+
+// Volume tests run against either a local test server or live on EC2.
+
+func (s *ServerTests) TestVolumes(c *C) {
+	vol1 := ec2.CreateVolume{
+		AvailZone:  "us-east-1b",
+		VolumeType: "standard",
+		VolumeSize: 20,
+	}
+	resp1, err := s.ec2.CreateVolume(vol1)
+	c.Assert(err, IsNil)
```

Comment:
I think you ought to add a deferred DeleteVolume right after creating the volume, in case the rest of the test fails.